crates/wasi-common/src/sys/windows/hostcalls_impl/misc.rs

```diff
@@ -77,12 +77,188 @@ pub(crate) fn clock_time_get(clock_id: wasi::__wasi_clockid_t) -> Result<wasi::_
     duration.as_nanos().try_into().map_err(Into::into)
 }
 
+fn stdin_nonempty() -> bool {
+    use std::io::Read;
+    std::io::stdin().bytes().peekable().peek().is_some()
+}
+
+fn make_read_event(event: &FdEventData, nbytes: Result<u64>) -> wasi::__wasi_event_t {
+    use crate::error::WasiErrno;
+    let error = nbytes.as_wasi_errno();
+    let nbytes = nbytes.unwrap_or_default();
+    wasi::__wasi_event_t {
+        userdata: event.userdata,
+        r#type: event.r#type,
+        error,
+        u: wasi::__wasi_event_u_t {
+            fd_readwrite: wasi::__wasi_event_fd_readwrite_t { nbytes, flags: 0 },
+        },
+    }
+}
+
+fn make_timeout_event(event: &FdEventData, timeout: &ClockEventData) -> wasi::__wasi_event_t {
+    wasi::__wasi_event_t {
+        userdata: timeout.userdata,
+        r#type: wasi::__WASI_EVENTTYPE_CLOCK,
+        error: wasi::__WASI_ERRNO_SUCCESS,
+        u: wasi::__wasi_event_u_t {
+            fd_readwrite: wasi::__wasi_event_fd_readwrite_t {
+                nbytes: 0,
+                flags: 0,
+            },
+        },
+    }
+}
+
 pub(crate) fn poll_oneoff(
     timeout: Option<ClockEventData>,
     fd_events: Vec<FdEventData>,
     events: &mut Vec<wasi::__wasi_event_t>,
-) -> Result<Vec<wasi::__wasi_event_t>> {
-    unimplemented!("poll_oneoff")
+) -> Result<()> {
+    use crate::fdentry::Descriptor;
+    use std::fs::Metadata;
+    use std::thread;
+
+    let timeout_duration = timeout
+        .map(|t| t.delay.try_into().map(Duration::from_nanos))
+        .transpose()?;
+
+    // With no events to listen, poll_oneoff just becomes a sleep.
+    if fd_events.is_empty() {
+        match timeout_duration {
+            Some(t) => {
+                thread::sleep(t);
+                return Ok(());
+            }
+            None => {
+                // The thread is not guanteed to remain parked forever, so we need to loop
+                loop {
+                    thread::park();
```

Comment:
Previously we were returning `Ok(())` for `fd_events.is_empty() && timeout.is_none()`.  I'm confused as to what this is attempting to do.

Reply:
`poll` invoked with `nfds = 0`, `timeout = -1` seems to be an infinite sleep (on my system, at the very least)

Reply:
This is now differing from the Unix implementation, which remains:

```rust
if fd_events.is_empty() && timeout.is_none() {
    return Ok(());
}
```

I still think we should be returning `Ok(())` from the Windows implementation rather than implementing an infinite wait.

@sunfishcode should we codify this in WASI?

Reply:
I think the Unix implementation is wrong here. If `timeout` is `None`, it means the WASI application didn't include a clock event subscription. If the application also didn't include any files to watch, it should block forever. [WASI libc's implementation of pause currently depends on this](https://github.com/CraneStation/wasi-libc/blob/master/libc-bottom-half/sources/pause.c#L8), and it appears it doesn't work in Wasmtime on Unix right now as a result of this.

Reply:
I'll fiz the Unix version in a subsequent PR.

Reply:
Actually, musl uses `ppoll` unless a dedicated syscall is available, so wasi-libc is doing nothing unusual, see https://git.musl-libc.org/cgit/musl/tree/src/unistd/pause.c

Reply:
This seems rather broken to codify in WASI given 1) there's no concept of signals anymore and 2) the Windows implementation will literally enter a loop with no mechanism for continuing.

I'd prefer we did not implement this for Windows at the very least.

Reply:
Also, it might be worth considering whether or not having Wasm code responding to host process signals is a violation of sandboxing (I could certainly be convinced it is).

Reply:
That's a good point. I also agree that we shouldn't automatically translate a host signal to a guest wake-up-`pause` event.

The main other use for `pause` I know about that doesn't involve signals is to hang a program after a bug is detected to allow someone to attach a debugger and inspect the program state. But I guess you can still do that by doing a `poll` with a big timeout.

So I agree that we can change this to be an error, presumably `EINVAL`, so that if someone does this we won't silently succeed.

Reply:
So do you suggest we throw an error if `thread::park` is spuriously woken up?

Reply:
This is probably not the last word on this topic, but for now, I think the answer is that we don't `thread::park` in the case where there's nothing to wait for. We just return `EINVAL` immediately in that case.

Reply:
On the other hand, empty poll is something that I wouldn't expect not to work when writing portable code. Maybe using `unimplemented` would be a better option, indicating that we do intend to support empty poll in the future?

Reply:
In this moment, I find @peterhuene's reasoning convincing. Portable code using `pause` is probably assuming it has signal handlers which can interrupt the `pause`, so it's already broken to some degree.

I opened https://github.com/WebAssembly/WASI/pull/193 to propose changing the spec to document that an empty polll returns einval. If there's consensus for this, I expect we'd remove `pause` from wasi-libc, so that portable code doesn't run into this by accident.